downstairs/src/region.rs

```diff
@@ -965,6 +1106,162 @@ impl Region {
+            .await
+            .unwrap()
+            .into_inner()
+            .files;
+
+        println!("Found repair files: {:?}", repair_files);
+
+        // The repair file list should always contain the extent data
+        // file itself, and the .db file (metadata) for that extent.
+        // Missing these means the repair will not succeed.
+        let filename = extent_file_name(eid as u32, None);
+        if !repair_files.contains(&filename) {
+            // XXX Panic now, but this will eventually bail and abort
+            // the repair and let the upper layers handle it.
+            panic!("Repair file list missing data file {}", filename);
+        }
+        let extent_copy =
+            extent.create_copy_file(copy_dir.clone(), None).unwrap();
+        let repair_stream = repair_server.get_extent(eid as u32).await.unwrap();
+        save_stream_to_file(extent_copy, repair_stream.into_inner()).await?;
+
+        let filename = extent_file_name(eid as u32, Some("db"));
+        if !repair_files.contains(&filename) {
+            // XXX Panic now, but this will eventually bail and abort
+            // the repair and let the upper layers handle it.
+            panic!("Repair file list missing db file {}", filename);
+        }
+        let extent_db = extent
+            .create_copy_file(copy_dir.clone(), Some("db"))
+            .unwrap();
+        let repair_stream = repair_server.get_db(eid as u32).await.unwrap();
+        save_stream_to_file(extent_db, repair_stream.into_inner()).await?;
+
+        // These next two are optional.
+        let filename = extent_file_name(eid as u32, Some("db-shm"));
+        if repair_files.contains(&filename) {
+            let extent_shm = extent
+                .create_copy_file(copy_dir.clone(), Some("db-shm"))
+                .unwrap();
+            let repair_stream =
+                repair_server.get_shm(eid as u32).await.unwrap();
+            save_stream_to_file(extent_shm, repair_stream.into_inner()).await?;
+        } else if repair_files.len() != 2 {
+            panic!("Unknown files on repair list: {:?}", repair_files);
+        }
+
+        let filename = extent_file_name(eid as u32, Some("db-wal"));
+        if repair_files.contains(&filename) {
+            let extent_wal = extent
+                .create_copy_file(copy_dir.clone(), Some("db-wal"))
+                .unwrap();
+            let repair_stream =
+                repair_server.get_wal(eid as u32).await.unwrap();
+            save_stream_to_file(extent_wal, repair_stream.into_inner()).await?;
+        }
+        if repair_files.len() > 4 {
+            panic!("Unknown extra files on repair list: {:?}", repair_files);
+        }
+
+        // After we have all files: move the repair dir.
+        println!(
+            "Repair files downloaded, move directory {:?} to {:?}",
+            copy_dir, rd
+        );
+
+        // XXX fsync the parent directory (the extent dir)
```

Comment:
https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_all should work here